src/JBDistributor.sol

```diff
@@ -1,208 +1,286 @@
+    //*********************************************************************//
+    // ---------------------- external transactions ---------------------- //
+    //*********************************************************************// 
+
+    /// @notice Claims tokens and beings vesting.
+    /// @param _tokenIds the ids to claim rewards for
+    /// @param _tokens the tokens to claim
     function beginVesting(uint256[] calldata _tokenIds, IERC20[] calldata _tokens) external {
+
+        // Keep a reference to the current round.
         uint256 _currentRound = currentRound();
+
+        // Keep a reference to the total staked amount at the curren round.
         uint256 _totalStakeAmount = _totalStake(roundStartBlock(_currentRound));
 
-        // Calculate the round in which the current rewards will release
+        // Calculate the round at which the current rewards will be release.
         uint256 _vestingReleaseRound = _currentRound + vestingRounds;
 
-        for (uint256 _i; _i < _tokens.length;) {
-            IERC20 _token = _tokens[_i];
+        // Keep a reference to the number of tokens that vesting is starting for.
+        uint256 _numberOfTokens = _tokens.length;
+
+        // Keep a reference to the token being iterated on.
+        IERC20 _token;
+
+        // Loop through each token of which vested is begining.
+        for (uint256 _i; _i < _numberOfTokens;) {
+            // Set the token.
+            _token = _tokens[_i];
 
             // Scoped to prevent stack too deep
             uint256 _distributable;
             {
-                // Check if a snapshot has been done of the token balance yet
-                // no: take snapshot
-                TokenState memory _state = _snapshotToken(_token);
-                _distributable = _state.balance - _state.vestingAmount;
+                // Take a snapshot of the token balance if it hasn't been taken already.
+                TokenSnapshotData memory _snapshot = _takeSnapshotOf(_token);
+                _distributable = _snapshot.balance - _snapshot.vestingAmount;
             }
 
+            // Keep a reference to the total amount vested.
             uint256 _totalVestingAmount;
-            for (uint256 _j; _j < _tokenIds.length;) {
-                // TODO: Make sure sender owns the token
-                // TODO: Do we even need to check for ownership? The owner can always choose to not collect
-                // TODO: Cache '_tokenStake' call
-                // Get the staked amount for the token
-                uint256 _tokenAmount = _distributable * _tokenStake(_tokenIds[_j]) / _totalStakeAmount;
-
-                // Check if this token was already claimed (check might not be needed)
-                if (tokenVesting[_tokenIds[_j]][_vestingReleaseRound][_token] != 0) {
-                    revert AlreadyVesting();
-                }
+
+            // Keep a reference to the number of token IDs.
+            uint256 _numberOfTokenIds = _tokenIds.length;
```

Comment:
nit> cost is the same (with an array in calldata, optimizier on and solc 0.8.20), *but* since we don't reuse numberOfTokenIds, current version is actually 6 units more expensive (to push/pop _numberOfTokenIds into stack)

<img width="559" alt="image" src="https://github.com/Bananapus/bananapus-distributor/assets/83670532/109e0a7f-92ef-4370-aa7d-f15dd79ef03e">

Reply:
meh, was intended for the higher level loop, but both are similar (calldata array, not reusing the cached length beside the loop test)